### PR TITLE
Added IsFinite Support for the tflite.

### PR DESCRIPTION
**Pull request by amitsrivastava78**
This is part of #21526

### CODE REVIEW

**Comment by amitsrivastava78** [comment]
@renjie-liu , can you please review the PR and provide your feedback.

Regards
Amit

**Review by renjie-liu — COMMENTED** [review]
can you add generate_zip_tests as well?

**Comment by amitsrivastava78** [comment]
@renjie-liu  thanks for the feedback i have made the changes as per your suggestion, kindly check.


Regards
Amit

**Comment by amitsrivastava78** [comment]
@renjie-liu  , rebased and pushed again.

Regards
Amit

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (renjie-liu) [inline tensorflow/lite/testing/generate_examples_lib.py]: instead of using _make_elementwise_tests, I think you need change it a little bit, since the generate test case will never populate infinite data, the default range is -100 to 100,  see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/testing/generate_examples_lib.py#L271-L288

Comment (amitsrivastava78) [reply tensorflow/lite/testing/generate_examples_lib.py]: @renjie-liu , thanks for pointing this out, i have updated the code as per your suggestion, which randomises the nan placement. Hope this is fine, kindly have a look.

Regards
Amit

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (renjie-liu) [inline tensorflow/lite/testing/generate_examples_lib.py]: maybe instead of adding this one, you can change the create_tensor_data to support infinity, nan, etc.

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (renjie-liu) [inline tensorflow/lite/testing/generate_examples_lib.py]: or you can just build a graph like:

input1 = tf.constant(shape=..., dtype=...)
input2 = tf.constant(shape=..., dtype=...)
out = tf.is_finite(input1 / input2)

then when you feed in values, input2 should contain some zeros.

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (amitsrivastava78) [reply tensorflow/lite/testing/generate_examples_lib.py]: @renjie-liu , sure i have modified the same function and added support for infinity

Regards
Amit

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (amitsrivastava78) [reply tensorflow/lite/testing/generate_examples_lib.py]: @renjie-liu , I have taken your first suggestion and modified the code as per that.

Regards
Amit

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (renjie-liu) [inline tensorflow/lite/testing/generate_examples_lib.py]: are you sure this is correct? when it's >1 dimensions, seems this will be a broadcast operation.

and also can you format it as well?

Comment (amitsrivastava78) [reply tensorflow/lite/testing/generate_examples_lib.py]: @renjie-liu , I have updated the code and replaced the logic for > 1 dimension as well, kindly have a look.

Regards
Amit

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (renjie-liu) [inline tensorflow/lite/testing/generate_examples_lib.py]: can this be extended to other types? if not, can you add a comment?

File: tensorflow/lite/kernels/elementwise_test.cc
Comment (renjie-liu) [inline tensorflow/lite/kernels/elementwise_test.cc]: nit: since it's test against finite, better include infinity data as well.

**Review by renjie-liu — COMMENTED** [review]
lgtm, but since it involves in schema change, please wait for another review as well.

File: tensorflow/lite/testing/generate_examples_lib.py
Comment (amitsrivastava78) [reply tensorflow/lite/testing/generate_examples_lib.py]: @renjie-liu ,Sure i have added a comment here.

Regards
Amit

File: tensorflow/lite/kernels/elementwise_test.cc
Comment (amitsrivastava78) [reply tensorflow/lite/kernels/elementwise_test.cc]: @renjie-liu , i have updated as per your suggestion.

Regards
Amit

**Comment by amitsrivastava78** [comment]
> lgtm, but since it involves in schema change, please wait for another review as well.

@renjie-liu , thanks alot for the review, highly appreciate your comments. I have made all the modifications, will wait as suggested by you.

Regards
Amit

**Comment by gbaned** [comment]
@amitsrivastava78 could you please resolve the conflicts? Thank you !

**Comment by amitsrivastava78** [comment]
@gbaned , thanks for pointing this out, i have resolved the conflicts, kindly check.

Regards
Amit

**Comment by gbaned** [comment]
@renjie-liu Can you please review this PR ?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle, @aselle, @renjie-liu, @miaout17, @renjie-liu, @renjie-liu, @renjie-liu, @renjie-liu: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Review by renjie-liu — DISMISSED** [review]
approve, please wait for one more lgtm

**Comment by amitsrivastava78** [comment]
@renjie-liu , thanks for approving the PR, there was one small sanity failure due line 271 of file tensorflow/lite/testing/generate_examples_lib.py due to column exceeding 80 characters, this is fixed now, can you, please re approve the PR.

Regards
Amit

**Review by renjie-liu — DISMISSED** [review]

**Comment by amitsrivastava78** [comment]
@renjie-liu , i have resolved the merge conflicts, can you please check and approve.


Regards
Amit

**Comment by amitsrivastava78** [comment]
@renjie-liu , i have resolved the merge conflicts, can you please check and re approve.

Regards
Amit

**Review by renjie-liu — APPROVED** [review]

**Comment by jdduke** [comment]
Which model(s) are you using this operator? Our general policy for builtin op additions is:

1) Prefer using select TF ops for one-off cases.
2) Provide a custom op if needed for less common ops if some amount of optimization is required.
3) Add op as builtin if there is significant demand from multiple clients and models.

I'm not sure this case satisfies (3).

**Comment by gbaned** [comment]
@amitsrivastava78 Could you please resolve the conflicts? Thanks!

**Comment by jdduke** [comment]
Can you respond per [this comment](https://github.com/tensorflow/tensorflow/pull/27710#issuecomment-513290343)?

**Comment by amitsrivastava78** [comment]
@jdduke , sorry for the late response, the model that i can see which uses this operator are t2t  transformer model (machine translation models) , there are other models which are based on machine translation (T2T) which also contains this operator.
 Hope this is ok.If you are ok then i will rebase the code and upstream again.

Regards
Amit

**Comment by jdduke** [comment]
Thanks, can you provide a link to the model(s)? I want to confirm that the operator is part of the inference subgraph, not just the training subgraph.

**Comment by amitsrivastava78** [comment]
@jdduke , please find the link for the model

wget -qO- --show-progress https://dl.dropbox.com/s/efv2gmq5hu3np43/log.tar.gz | tar xz


Regards
Amit

**Comment by amitsrivastava78** [comment]
But it is this operator is used in gradient calculations

Regards 
Amit

**Comment by gbaned** [comment]
@amitsrivastava78 Could you please resolve the conflicts? Thanks!

**Comment by jdduke** [comment]
> But it is this operator is used in gradient calculations

Right, which we shouldn't need for inference with TFLite. Let's close this until we find a compelling inference need. Thanks!

**Pull request closed without merging**
